src/assets/ts/session.ts

```diff
@@ -1335,6 +1335,33 @@ export default class Session extends EventEmitter {
 
     this.emit('scroll', numlines);
   }
+
+  public async getTextRecusive(path: Path) {
+    let result: string[] = [];
+
+    if (await this.document.collapsed(path.row)) {
+      throw new Error('Some blocks is folded!');
```

Comment:
nit: "Some blocks are folded!"

Reply:
optional:  I would probably make the error a return value

    let err: string | null = null;

    if (await this.document.collapsed(path.row)) {
      err = 'Some blocks are folded!';
    }

    ...


      for (let [childErr, childResult] of resultChildren) {
        if (childErr !== null) { err = childErr; }
        result.push(childResult);
      }

....
      return err, result.join('\n')

Reply:
another thing to consider is to make the return value here an array, and do the `result.join('\n')` in `visual-line-join` instead

Reply:
> nit: "Some blocks are folded!"
You are right. Sorry, English is not my native language.

Reply:
no problem!

Reply:
> optional: I would probably make the error a return value
About this: I can't. I tried many options, and failed to return the value with an error.

for (let [childErr, childResult] of resultChildren) {

Type 'string' is not an array type or a string type. Use compiler option '--downlevelIteration' to allow iterating of iterators.ts(2569)

Can you help with it?

Reply:
in `tsconfig.json`, add    ```"downlevelIteration": true,``` and then rerun `npm start`, and i think it should work

Reply:
I can't get it work.

 ```
 public async getTextRecusive(path: Path) {
    let result: string[] = [];
    let err: string | null = null;

    if (await this.document.collapsed(path.row)) {
      err = 'Some blocks are folded!';
    }

    const text = await this.document.getText(path.row);
    result.push(text);

    if (await this.document.hasChildren(path.row)) {
      let children = await this.document.getChildren(path);

      const resultChildren = await Promise.all(
        children.map(async (childrenPath) => {
          return await this.getTextRecusive(childrenPath);
        })
      );

      for (let [childErr, childResult] of resultChildren) {
        if (childErr !== null) { err = childErr; }
        result.push(childResult);
      }
    }

    return err, result.join('\n');
  }
}
```

err from child recursive function does not return.

Reply:
so it compiles, but `err = childErr` doesn't work correctly?

Reply:
Yes, it compiles but in that case childErr is first letter of line, childResult second.

Reply:
you need to also change `visual-line-join`, right?

Reply:
I don't understand how it works. Can I somewhere seethe solution to a similar problem? Google doesn't help.

Reply:
thanks, i'll merge this and look into it myself!

Reply:
made it work here, if you're curious: https://github.com/WuTheFWasThat/vimflowy/commit/df958af1c0450408caa37a3b2710b46239cd9f68

Reply:
> made it work here, if you're curious: [df958af](https://github.com/WuTheFWasThat/vimflowy/commit/df958af1c0450408caa37a3b2710b46239cd9f68)

Oh, thanks. I would never code like this.